### DISCUSSION TITLE
Having __init__.py file in the root fails to import modules in the using --onefile

### DISCUSSION THREAD

**Discussion by marcinjrojek** [pyinstaller/pyinstaller#6728] [Help]
I'm having the same issue as here https://stackoverflow.com/questions/69978685/pyinstaller-fails-to-find-packages-when-using-onefile
 and I'm suspecting it might be a bug because my `compile.bat` script stopped working after upgrading to `pyinstaller==4.10`.

Removing __init__.py is not an option because "If some tools like pytest and unittest still needs the _init_.py file then it's not totally correct to say it's not necessary for packages anymore." as explained here.
https://stackoverflow.com/questions/59837593/is-the-init-py-really-not-necessary-for-python-3-7-packages

**Comment by bwoodsend** [comment]
You're probably on the wrong end of [this quirk](https://github.com/pyinstaller/pyinstaller/issues/6713#issuecomment-1083684058). That said, it will only be a problem if you're already doing something invalid. Is you're project layout the same as that 1st StackOverflow link because that is invalid - that top level `__init__.py` does not belong there. By executing `myscript.py` as `__main__`, you add that folder to Python's module search path which means that `mylib` is a Python package but that top level directory certainly isn't and should therefore not contain an `__init__.py`.

**Reply by bwoodsend** [reply]
The correct way to organise that project would be to move `myscript.py` to `mylib/__main__.py` then you can invoke it normally via `python -m mylib` and build it with `pyinstaller mylib/__main__.py`.

**Reply by marcinjrojek** [reply]
Thanks for a very quick reply.
I've put this MCVE together. This is the structure of my app: [click.](https://github.com/marv2018/my_app/)
Currently, my project layout looks like the `run_app1.py/build_app1.bat`  ie. I run `run_app1.py` as `__main__`.
Removing `my_app\src\my_app\__init__.py` fixes `ModuleNotFoundError` when building with pyinstaller. 
However `my_app` is really a package therefore it needs `__init__.py`, doesn't it ?...

As for pytest, I'm able to execute(from command prompt) `python -m pytest test/test_abc`

**Reply by bwoodsend** [reply]
When you use `python src/run_app2.py`, `src` is added to `sys.path` so that any folder inside `src` (such as `my_app`) is an `import`-able package and any other `.py` file (of which there are none) become `import`-able modules.

When you use `python src/my_app/run_app1.py` then `src/my_app` is added to `sys.path` so that any folder inside `src/my_app`, such as `resources`, is a package and any other `.py` files in there (`meta_info.py` and even the `__init__.py`) become modules. `my_app` itself is now just a folder.

What makes a package different from a random folder of `.py` files is not whether or not there's an `__init__.py` in there - it's if its parent directory is in `sys.path`.

So in your example repo, you can either move `src/my_app/run_app1.py` to `src/run_app1.py` like `run_app2.py` or can add `src` to `sys.path` then invoke as `python -m my_app.run_app1`. To add to `sys.path` you can either use the temporary dirty solution of setting `PYTHONPATH`:
```
set PYTHONPATH=C:\path\to\my_app\src
```
Or the proper, more permanent solution is to add a `setup.py` at the root of your repository:
```python
# setup.py
from setuptools import setup, find_packages

setup(
    name="my_app",
    install_requires=["PyQt5"],
    packages=find_packages(include=["my_app", "my_app.*"], where="src"),
    package_dir={"": "src"},
    package_data={"my_app": ["resources/view/*.ui"]},
    version="0.1.0",
)
```
Then you can install your package with `pip`:
```
pip install -e .
```
Then run `python -m my_app.run_app1` anywhere you like. (Note, you'd first have to fix these incorrect imports.)
```diff
diff --git i/src/my_app/resources/view/main_window.py w/src/my_app/resources/view/main_window.py
index aa0c440..d67a1c7 100644
--- i/src/my_app/resources/view/main_window.py
+++ w/src/my_app/resources/view/main_window.py
@@ -1,5 +1,5 @@
 from PyQt5.QtWidgets import QMainWindow
-from resources.view.main_window_ui import Ui_MainWindow
+from my_app.resources.view.main_window_ui import Ui_MainWindow
 
 class MainWindow(QMainWindow):
     def __init__(self, parent=None):
diff --git i/src/my_app/run_app1.py w/src/my_app/run_app1.py
index f0c868b..12f7210 100644
--- i/src/my_app/run_app1.py
+++ w/src/my_app/run_app1.py
@@ -1,8 +1,8 @@
 import sys
 
 from PyQt5.QtWidgets import QApplication
-from resources.view.main_window import MainWindow
-from meta_info import __app_version__, __app_name__
+from my_app.resources.view.main_window import MainWindow
+from my_app.meta_info import __app_version__, __app_name__
 
 class App(QApplication):
     def __init__(self, sys_argv):
```

Once you've done that, you can delete the `requirements.txt` and tell people to install your app with one line:
```
pip install git+https://github.com/marv2018/my_app.git
```
